core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -197,16 +197,41 @@ public int unfinishedRequests() {
     }
 
     @Override
-    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
-        super.channelUnregistered(ctx);
-        unfinishedResponses.keySet().forEach(StreamMessage::abort);
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        // Give the unfinished streaming responses a chance to close themselves before we abort them,
+        // so that successful responses are not aborted due to a race condition like the following:
+        //
+        // 1) A publisher of a response stream sends the complete response
+        //    but does not call StreamWriter.close() just yet.
+        // 2) An HTTP/1 client receives the complete response and closes the connection, which is totally fine.
+        // 3) The response stream is aborted once the server detects the disconnection.
+        // 4) The publisher calls StreamWriter.close() but it's aborted already.
+        //
+        // To reduce the chance of such situation, we wait a little bit before aborting unfinished responses.
+
+        switch (protocol) {
+            case H1C:
+            case H1:
+                // XXX(trustin): How much time is 'a little bit'?
+                ctx.channel().eventLoop().schedule(this::cleanup, 1, TimeUnit.SECONDS);
+                break;
+            default:
+                // HTTP/2 is unaffected by this issue because a client is expected to wait for a frame with
+                // endOfStream set.
+                cleanup();
+        }
     }
 
-    @Override
-    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+    private void cleanup() {
         if (responseEncoder != null) {
             responseEncoder.close();
         }
+
+        unfinishedResponses.keySet().forEach(res -> {
+            if (!res.isComplete()) {
```

Comment:
Might be mistaken but I thought our semantics allowed aborting already completed streams (since completion is asynchronous I thought it's considered best effort vs other completion events)

Reply:
Yeah, it's probably redundant. Let me run the tests again without this guard.

Reply:
Everything's fine without the guard. Removed.